### ISSUE TITLE
코드

### ISSUE THREAD

**Issue by sh625-625** [sh625-625/Project1#5]
import os
import re
import sys
import io
import pytesseract
import pandas as pd
from PIL import Image
from openpyxl import load_workbook

# 콘솔 한글 깨짐 방지 (Windows)
sys.stdout = io.TextIOWrapper(sys.stdout.detach(), encoding='utf-8')

# Tesseract 경로 및 설정 (Windows)
pytesseract.pytesseract.tesseract_cmd = r"C:\Program Files\Tesseract-OCR\tesseract.exe"
custom_config = r'--oem 3 --psm 6'

# 입력/출력 경로
img_path    = r"c:\yoonsung\financial ledger automation\financial ledger automation\receipt.jpg"
output_path = r"c:\yoonsung\financial ledger automation\financial ledger automation\가계부_자동완성.xlsx"

# 1. 이미지 열기
pil_img = Image.open(img_path)

# 2. OCR 수행 및 라인 전처리
raw = pytesseract.image_to_string(pil_img, lang="kor+eng", config=custom_config)
lines = []
for ln in raw.splitlines():
    s = ln.strip()
    if not s:
        continue
    # "9, 300" → "9,300"
    s = re.sub(r",\s+", ",", s)
    # 특수문자(*) _ ( ) 제거
    s = re.sub(r"[\*\_\(\)]", "", s)
    lines.append(s)

# 3. 상품명/바코드/수량/가격 추출
entries = []
prev_name = None

for ln in lines:
    # (1) 상품명 줄: 맨 앞에 3자리 숫자 코드 + 텍스트 포함
    if re.match(r"^\d{3}", ln) and re.search(r"[가-힣A-Za-z]", ln):
        # 코드(003 등) 제거
        raw_name = re.sub(r"^\d{3}\s*", "", ln).strip()
        # 상품명은 공백 전까지
        prev_name = raw_name.split()[0]
        continue

    # (2) 숫자 줄 & prev_name 있을 때
    if prev_name and re.search(r"\d", ln):
        raw_nums = re.findall(r"\d{1,3}(?:,\d{3})*|\d+", ln)
        nums = [n.replace(",", "") for n in raw_nums]

        barcode = ""
        qty = None
        price = None

        # a) 첫번째 10자리 이상 토큰 → 바코드
        for i, token in enumerate(nums):
            if len(token) >= 10:
                barcode = token
                # b) 그 다음 토큰은 무시, 이후 한 자리 수량 찾아서
                for j in range(i + 2, len(nums)):
                    if len(nums[j]) == 1:
                        qty = int(nums[j])
                        # c) 수량 다음 토큰을 가격으로
                        if j + 1 < len(nums):
                            price = int(nums[j + 1])
                        break
                break

        # 바코드 못 찾으면, 마지막 두 토큰을 qty/price로
        if not barcode and len(nums) >= 2:
            qty = int(nums[-2])
            price = int(nums[-1])

        # qty/price 기본값
        if qty is None:
            qty = 1
        if price is None and nums:
            price = int(nums[-1])

        entries.append({
            "상품명": prev_name,
            "바코드": barcode,
            "수량":   qty,
            "가격":   price
        })

    prev_name = None

# 4. DataFrame 생성 및 상품명 최종 정리
df = pd.DataFrame(entries)
# 숫자·특수문자 제거: 한글·영문만 남김
df["상품명"] = df["상품명"].str.replace(r"[^가-힣A-Za-z]", "", regex=True)

# 5. 출력 폴더 생성
os.makedirs(os.path.dirname(output_path), exist_ok=True)

# 6. 엑셀 저장
df.to_excel(output_path, index=False, sheet_name="가계부")

# 7. 열 너비 조정
wb = load_workbook(output_path)
ws = wb["가계부"]
ws.column_dimensions["A"].width = 30  # 상품명
ws.column_dimensions["B"].width = 20  # 바코드
ws.column_dimensions["C"].width = 8   # 수량
ws.column_dimensions["D"].width = 12  # 가격
wb.save(output_path)

print(f"완료: {len(df)}개 항목이 추출되어\n→ {output_path}에 저장되었습니다.")
